### PR TITLE
Update docs. Add design principles. 

### LINKED ISSUE #4182: Merge support matrix and maintainers list
Lots of overlap:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/getting-started-guides/README.md
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/MAINTAINERS.md
cc @erictune

### LINKED ISSUE #6133: Document API/system design principles and patterns
Related to #5476, #2003, #1007, #5620, and other issues.

The PR creation rate and number of contributors are both growing. In order to scale the effort, we need to better document the patterns we want additions to the system and API to follow, so that we can better parallelize the design and review effort. I'm filing this issue to solicit important principles and patterns to document. I volunteer to create the actual document and/or presentation slides.

What do you think isn't sufficiently covered by existing design docs and/or isn't sufficiently discoverable? What principles/patterns do you find yourself explaining over and over again? What properties would you be unhappy to have violated/degraded/changed by new PRs? 

I'm thinking more about reasonably general issues rather than implementation-oriented invariants that apply to very specific/narrow parts of the code. 

Examples (some areas, some principles):
* Declarative APIs, desired state and current state
* Level-based operation: Functionality must be level-based, meaning the system must operate correctly given the desired state and the current/observed state, regardless of how many intermediate state updates may have been missed. Edge-based behavior must be just an optimization.
* Status must be 100% reconstructable by observation. Any history kept must be just an optimization and not required for correct operation.
* Phase, Reason, Message, Conditions, etc. vs. a single state machine with lots of states
* Composable, a la carte functionality
* Transparent, no glass to break
* More generally, [Eric Raymond's 17 UNIX rules](https://en.wikipedia.org/wiki/Unix_philosophy#Eric_Raymond.E2.80.99s_17_Unix_Rules)
* Only the apiserver should communicate with etcd/store, and not other components (scheduler, kubelet, etc.).
* Compromising a single node shouldn't compromise the cluster (not that we've achieved that yet)
* Components should continue to do what they're told in the absence of new instructions.  
* Open world assumption: continually verify assumptions and gracefully adapt to external events and/or actors (example: we allow users to kill pods under control of a replication controller; it just replaces them)
* Don't assume a component's decisions will not be overridden or rejected, nor for the component to always understand why. For example, etcd may reject writes. Kubelet may reject pods. The scheduler may not be able to schedule pods.
* When to use labels vs. annotations vs. object references
* Watch vs. polling
* Bootstrapping-related principles

Let's brainstorm first, and argue and prioritize afterward.

@smarterclayton @derekwaynecarr @thockin @erictune @lavalamp

### PR DESCRIPTION
**Pull request by bgrant0607**
Fixes #6133. Fixes #4182.

DESIGN.md was horribly out of date, out of place, and overlapped with other docs, so I nuked it. I moved some bits of its content elsewhere. I see some bits still need to be updated (architecture.md), which I could do now, or in a subsequent PR, depending on how quickly this gets merged. :-)

cc @smarterclayton @lavalamp @thockin

### CODE REVIEW

**Comment by bgrant0607** [comment]
I'll update architecture.md now. It's egregiously out of date.

**Comment by bgrant0607** [comment]
Updated.

File: README.md [hunk 1]
```diff
@@ -18,7 +18,7 @@ Kubernetes is:
 * **extensible**: modular, pluggable, hookable, composable
 * **self-healing**: auto-placement, auto-restart, auto-replication
 
-Kubernetes builds upon a decade and a half of experience at Google running production workloads at scale, combined with best-of-breed ideas and practices from the community.
+Kubernetes builds upon a [decade and a half of experience at Google running production workloads at scale](https://research.google.com/pubs/pub43438.html), combined with best-of-breed ideas and practices from the community.
```
Comment (lavalamp) [inline README.md]: Hm, maybe "decade and a half of experience running production workloads at Google scale" flows a tiny bit better?

File: docs/api-conventions.md [hunk 2]
```diff
@@ -1,11 +1,11 @@
...
 
-The general style of the Kubernetes API is RESTful - clients create, update, delete, or retrieve a description of an object via the standard HTTP verbs (POST, PUT, DELETE, and GET) - and those APIs preferentially accept and return JSON. Kubernetes also exposes additional endpoints for non-standard verbs and allows alternative content types. All of the JSON accepted and returned by the server has a schema, identified by the "kind" and "apiVersion" fields.
+The general style of the Kubernetes API is RESTful - clients create, update, delete, or retrieve a description of an object via the standard HTTP verbs (POST, PUT, DELETE, and GET) - and those APIs preferentially accept and return JSON. Kubernetes also exposes additional endpoints for non-standard verbs and allows alternative content types. All of the JSON accepted and returned by the server has a schema, identified by the "kind" and "apiVersion" fields. Where relevant HTTP header fields exist, they should mirror the content of JSON fields, but the information should not be represented only in the HTTP header.
```
Comment (lavalamp) [inline docs/api-conventions.md]: "allows alternative content types"-- is that true? Or are you referring to the proxying or something?

Comment (smarterclayton) [inline docs/api-conventions.md]: > On Apr 16, 2015, at 7:43 PM, Daniel Smith <notifications@github.com> wrote:
> 
> In docs/api-conventions.md:
> 
> >  
> >  The conventions of the [Kubernetes API](api.md) (and related APIs in the ecosystem) are intended to ease client development and ensure that configuration mechanisms can be implemented that work across a diverse set of use cases consistently.
> >  
> > -The general style of the Kubernetes API is RESTful - clients create, update, delete, or retrieve a description of an object via the standard HTTP verbs (POST, PUT, DELETE, and GET) - and those APIs preferentially accept and return JSON. Kubernetes also exposes additional endpoints for non-standard verbs and allows alternative content types. All of the JSON accepted and returned by the server has a schema, identified by the "kind" and "apiVersion" fields.
> > +The general style of the Kubernetes API is RESTful - clients create, update, delete, or retrieve a description of an object via the standard HTTP verbs (POST, PUT, DELETE, and GET) - and those APIs preferentially accept and return JSON. Kubernetes also exposes additional endpoints for non-standard verbs and allows alternative content types. All of the JSON accepted and returned by the server has a schema, identified by the "kind" and "apiVersion" fields. Where relevant HTTP header fields exist, they should mirror the content of JSON fields, but the information should not be represented only in the HTTP header.
> "allows alternative content types"-- is that true? Or are you referring to the proxying or something?
> 
Yeah, we have pod logs (text/plain)

File: README.md (hunk 1 above)
Comment (bgrant0607) [inline README.md]: That text is copied all over. I could fix it, but I'm not sure it's worth it right now. I'm focused on content.

**Comment by lavalamp** [comment]
LGTM

File: docs/api-conventions.md (hunk 2 above)
Comment (bgrant0607) [inline docs/api-conventions.md]: PATCH also has other content types.

**Pull request merged**